src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+            /// has not yet been disposed; otherwise, <see langword="null"/> if this reference to the underlying object
+            /// has already been disposed.</returns>
+            public ReferenceCountedDisposable<T> TryAddReference()
+            {
+                // This value can be set in Dispose. However, even if we have a race condition with Dispose where it's
+                // set to null in the middle of the call, the reference count will be incremented by this method if and
+                // only if a valid new reference was created. This is because the validity of this instance for creating
+                // new references is determined by _referenceCount.
+                var target = _instance;
+                Interlocked.MemoryBarrier();
+                var referenceCount = Volatile.Read(ref _referenceCount);
+
+                // Cannot use Interlocked.Increment because we need to latch the reference count when it reaches 0.
+                while (true)
+                {
+                    var currentValue = Volatile.Read(ref referenceCount.Value);
+                    if (currentValue == 0)
+                    {
+                        // The target is already disposed, and cannot be reused
+                        return null;
+                    }
+
+                    if (Interlocked.CompareExchange(ref referenceCount.Value, currentValue + 1, currentValue) == currentValue)
+                    {
+                        return new ReferenceCountedDisposable<T>(target, referenceCount);
+                    }
+                }
+            }
+
+            public void Dispose()
+            {
+                var referenceCount = Interlocked.Exchange(ref _referenceCount, null);
+                if (referenceCount == null)
+                {
+                    // Already disposed; allow multiple without error.
+                    return;
+                }
+
+                // This is the thread which will write to _instance, so this will act as an atomic read
+                var instance = _instance;
+
+                // Set the instance back to its default value, so in case someone forgets to dispose of one of the
+                // counted references the finalizer of the underlying disposable object will have a chance to clean up
+                // the unmanaged resources as soon as possible.
+                Interlocked.MemoryBarrier();
+                _instance = default(T);
+
+#if !DEBUG
+                var decrementedValue = Interlocked.Decrement(ref referenceCount.Value);
+                if (decrementedValue == 0)
+                {
+                    instance.Dispose();
+                }
+#else
+                while (true)
+                {
+                    var currentValue = Volatile.Read(ref referenceCount.Value);
+                    Debug.Assert(currentValue > 0, "Dispose should have protected itself against races.");
```

Comment:
My overall point is that this entire PR is extremely hard to review and verify correctness on.  There have been at least two locations that look wrong *from a local analysis*.  I haven't even been able to get to the point of understanding about how this all works together *non-locally* to validate correctness.  

Can we move to a simpler and safer approach that gives us the semantics we seem to want (deterministic release if all outstanding references are disposed), and then move to something more complex *if* data demonstrates that there is actually a concern of something like contention overhead.

These codepaths are only hit for creating streams/SourceTexts for syntax trees we've dumped and need to read back in, or for metadata references in VS.  It's unclear to me why this would be an area that necessitates things like Lock-Free.